### ISSUE TITLE
[18.0] Odoo from getting started to giving up

### ISSUE THREAD

**Issue by guowei1651** [odoo/odoo#190669]
### Odoo Version

- [ ] 16.0
- [ ] 17.0
- [X] 18.0
- [ ] Other (specify)

### 1. Value analysis

Analyze the value that Odoo can provide at this stage from several angles, and rate it here on a ten-point scale.

#### **1. Architecture design (score: -1)**
- Old design patterns

 The hierarchical structure developed in the last century has evolved through many rounds of evolution from the original display-logic-storage method to MVC, MVP, MVVC and other forms. Odoo is still using the most primitive method, and has evolved a DDD model that is more suitable for application development.

- Blurred module boundaries

 The logical boundaries are extremely confusing. The tools module includes business logic and basic tool capabilities. And it is very difficult to separate the logic in tools. http code and cron code are mixed together.

- Confusing dependencies

 During the review of the design process, I found many confusing calling relationships. The most notable feature is that you see the import statement at the end of the python script file, and there are many circular reference scenarios.

#### **2. Technology (score: 0)**

Use the technology stack of the last century for management and development:
1. **Server-side rendering**: RESTful and GraphQL technologies are not used.
2. **Extensive use of XML**: Toml and Yaml technologies are not used
3. **File storage**: minio, oss technology not used
4. **Caching Technology**: Redis technology is not used

#### **3. Scalable (score: 0)**

Plan and design the overall system through modules. This is a very good mode expansion method, but you can compare the expansion mode of K8s. The expansion model of K8s is carried out using data standards + events + processing.
- Extended isolation

 Extensibility in odoo is to load modules into your own framework. This will lead to the fact that if there is some malicious code in the module, it cannot be blocked at all. Coupled with the chaos of odoo itself, it is difficult to investigate if the content in odoo is incorrectly operated in the module.

- Extended standardization

 The standardization of odoo uses Model, Controller, etc. for standardization, but it has not reached the level of standard implementation for available life cycle management.

#### **4. Business support capability (score: 0)**

Business support capabilities: i18n, import and export, scheduled tasks, authentication and authorization, SMS, email, collection and payment, full-text search, BPM, session management, file upload, AI, etc. From these aspects, the capabilities are not enough to be used out of the box, and some are not even supported.

#### **5. Development capability support (score: 0)**

It was originally thought that Odoo only needs to do model definition and display definition. You can complete database generation, API generation, and basic function generation. Then you need to implement a code that requires a little business logic implementation. Equivalent to the implementation of Low-Code. As a result, it is impossible to achieve this level of development capabilities.

#### **6. Integration (score: 0)**

In terms of integration, Odoo is still solving the lowest level technical problems and does not provide support from the perspective of business capabilities and development capabilities. Therefore, the ERP functions provided to support integrated capabilities cannot form complete ERP capabilities.

#### in conclusion:
Odoo is a product of the last century, but it has not kept pace with the world. For example: the separation of front-end and back-end in microservices, the use of Cache's KV database, and Session's file storage. However, Odoo uses Python3, a language that continues to evolve with the times. It can only be said that Odoo's brain believes that Python3 is more important than following the ideological development of the times.

**Comment by orfisko** [comment]
**The added value of your post (score: -10)**
Your 'issue'  shows a clear lack of understanding that Odoo is a tanker that is difficult to divert to another course. Have you ever done IT in the real world? 

Btw, I strongly disagree with a lot of your remarks. Yes, the development experience is a nightmare, but in my company Odoo is bringing huge amounts of value (for free!).

**Comment by duong77476-viindoo** [comment]
I 've seen you are trying to rewritten odoo code in a way that help you understand only , i find that very hilarious, 
BTW agree with `orfisko`, Odoo helps many SME enterprise and has been studied in many university

**Comment by gdgellatly** [comment]
lol, what a knob

**Comment by pumppi** [comment]
I was there too...

But when I understood how to use Odoo's framework, it is wonderful. With all the hype technology changing 2-3 years, maintenance of all things would be hard.

And by the way trend is moving from microservices back to monoliths. Just follow up things in twitter.

**Comment by guowei1651** [comment]
Read Clean Architecture by Robert Martin.
